src/main/java/org/tugraz/sysds/runtime/instructions/fed/AggregateBinaryFEDInstruction.java

```diff
@@ -81,95 +86,295 @@ public void processInstruction(ExecutionContext ec) {
+		}
+		else // MM
+			federatedAggregateBinary(mo1, mo2, out);
 	}
 	
 	/**
 	 * Performs a federated binary aggregation (currently only MV and VM is supported).
-	 * 
-	 * @param mo1 the federated matrix object
-	 * @param mo2 the other matrix object (currently has to be a vector)
-	 * @param op the operation
-	 * @param swapParams if <code>this</code> and <code>other</code> should be swapped (other op this, instead of
-	 *                   this op other)
+	 *
+	 * @param mo1 the first matrix object
+	 * @param mo2 the other matrix object
+	 */
+	private void federatedAggregateBinary(MatrixObject mo1, MatrixObject mo2, MatrixObject out) {
+		boolean distributeCols = false;
+		// if distributeCols = true we distribute cols of mo2 and do a MV multiplications, otherwise we
+		// distribute rows of mo1 and do VM multiplications
+		if (mo1.isFederated() && mo2.isFederated()) {
+			// both are federated -> distribute smaller matrix
+			// TODO do more in depth checks like: how many federated workers, how big is the actual data we send and so on
+			// maybe once we track number of non zeros we could use that to get a better estimation of how much data
+			// will be requested?
+			distributeCols = mo2.getNumColumns() * mo2.getNumRows() < mo1.getNumColumns() * mo1.getNumRows();
+		}
+		else if (mo2.isFederated() && !mo1.isFederated()) {
+			// Distribute mo1 which is not federated
+			distributeCols = true;
+		}
+		// TODO performance if both matrices are federated
+		Map<FederatedRange, FederatedData> mapping = distributeCols ? mo1.getFedMapping() : mo2.getFedMapping();
+		MatrixBlock matrixBlock = distributeCols ? mo2.acquireRead() : mo1.acquireRead();
+		ExecutorService pool = CommonThreadPool.get(mapping.size());
+		ArrayList<Pair<FederatedRange, MatrixBlock>> results = new ArrayList<>();
+		ArrayList<FederatedMMTask> tasks = new ArrayList<>();
+		for (Map.Entry<FederatedRange, FederatedData> fedMap : mapping.entrySet()) {
+			// this resultPair will contain both position of partial result and the partial result itself of the operations
+			MutablePair<FederatedRange, MatrixBlock> resultPair = new MutablePair<>();
+			// they all get references to the real block, the task slices out the needed part and does the
+			// multiplication, therefore they can share the object since we use it immutably
+			tasks.add(new FederatedMMTask(fedMap.getKey(), fedMap.getValue(), resultPair, matrixBlock, distributeCols));
+			results.add(resultPair);
+		}
+		CommonThreadPool.invokeAndShutdown(pool, tasks);
+		if (distributeCols)
+			mo2.release();
+		else
+			mo1.release();
+		// combine results
+		out.acquireModify(combinePartialMMResults(results, mo1.getNumRows(), mo2.getNumColumns()));
+		out.release();
+	}
+	
+	private MatrixBlock combinePartialMMResults(ArrayList<Pair<FederatedRange, MatrixBlock>> results, long rows,
+			long cols) {
+		// TODO support large blocks with > int size
+		MatrixBlock resultBlock = new MatrixBlock((int) rows, (int) cols, false);
```

Comment:
does it have to be long inputs for this function? We risk overflows here. and yes i have seen the todo.

Reply:
I will switch to int, but will keep the todo

Reply:
btw I still have to cast at the position of the call. Where would you prefer the cast?